Encrypted rootfs release

tomzy-0 opened the pull request:
Gather all tests that were implemented while working on Encrypted rootfs based on TPM

PLangowski committed:
dasharo-security: tpm-support: add TPM Clear procedure test

Signed-off-by: Pawel Langowski <pawel.langowski@3mdeb.com>

Commit:
README: fix syntax

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
.gitignore: add qemu-data used in tests

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
scripts: qemu-run: fix typo in warning comment

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

m-iwanicki committed:
lib: tpm-support-lib.robot: Return List of strings

containing concatenated keys and pairs instead of dictionary

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
keywords: Split Boot Operating System keyword between AMI and Dasharo

Add more variables to platform config.

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
lib: tpm-support: split files into ami/dasharo

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
dasharo-security: tpm-support: fix TPM001.004 documentation

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
lib/bios/menus: Remove unused arguments from keyword Save Changes And Reset

Update other files that were using arguments

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
lib: tpm-support: Add keyword that checks TPM version fix option merging

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
dasharo-security/tpm-support.robot: remove duplicate test

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

Commit:
lib: QemuMonitor: set ROBOT_LIBRARY_SCOPE to SUITE

To use Power On keyword in Setup Suite section, we need to set
ROBOT_LIBRARY_SCOPE variable.
See: http://robotframework.org/robotframework/latest/RobotFrameworkUserGuide.html#test-library-scope

This is necessary to use KW from this lib in setup that runs
ansible-playbooks on target.

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
scripts: qemu-run: Add DRIVE_PATH, which allows adding virtual drives

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
platform-configs: qemu: add variable used in tests with DTS and UEFI SB enabled

see: https://github.com/Wind-River/meta-secure-core/tree/master/meta-efi-secure-boot#quick-start-for-the-first-boot

This is a procedure integrated in DTS for UEFI Secure Boot enabled
build.

see: https://github.com/Dasharo/meta-dts/blob/main/kas-uefi-sb.yml

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
lib: Split menus and secure-boot into ami/dasharo files

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
platform-configs: resource needed bios and secure-boot robot files

For those platforms we are using Dasharo firmware

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
variables: add MPL-PIP4 platform

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
platform-configs: mpl-pip4: add configuration for MPL-PIP4 device

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

m-iwanicki committed:
lib: dts-lib: shorten pikvm DTS name due to AMI limit

Don't wait for DTS to boot when certificate_provisioning == True
when using pikvm connection method

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

Commit:
lib: ansible: add KW to restore connection method

Needed as sometimes RF change connection to SSH

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
variables: use IP address for MPL sonoff

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
dasharo-security: tpm-support: make sure to boot Linux before running tests

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

m-iwanicki committed:
lib: tpm-support: remove unused files

They were moved to lib/dasharo/tpm-support-lib-dasharo.* in previous
commits

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
dasharo-security: tpm-support: remove tianocore from keyword

Fail when trying to add not implemented boot option. Remove unused
arguments from Save Changes And Reset.

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

m-iwanicki committed:
lib: secure-boot-dasharo: change default value

of sb_menu in Enter Key Management And... to empty list instead of
empty string.

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

tym2k1 committed:
dasharo-security/tpm2-commands.robot: added missing unseal check

Signed-off-by: Tymoteusz <tymbur@gmail.com>

Commit:
lib: bios: menus-dasharo: fix typo for checkpoint

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

m-iwanicki committed:
lib: secure-boot-dasharo: wait for sb keys to reset before finishing kw

Signed-off-by: Michał Iwanicki <michal.iwanicki@3mdeb.com>

Commit:
dasharo-security: secure-boot: add SBO009.001

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
dasharo-security: secure-boot: add SBO010.001-006

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
dasharo-security: secure-boot: add SBO011.001

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
dasharo-security: secure-boot: add SBO012.001

Add needed keywords in secure-boot-lib and fix skips in dasharo-security

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
Introduce ansible usage in osfv

* adds new packages to requirements.txt
* adds ansible-roles directory with tests example of usage
* set default value for ANSIBLE_CONFIG variable
* adds section in README about ansible usage
* adds Run Ansible Playbook On Supported Operating Systems keyword
* test ansible runs on secure-boot.robot

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
lib: secure-boot-lib: add keywords for UEFI SB tests in OS

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
dasharo-security: secure-boot: add SBO013.001-002 and SBO014.001

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
dasharo-security: secure-boot: add SBO015.001

Expand Enroll Secure Boot Keys In OS KW, add fixes to SBO014.001
and SBO012.001

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
Introduce sb-img-wrapper script

The script generates ISO images with efi executables used in UEFI Secure
Boot tests (where we test different scenarios of good/bad signed files
or with differnet cryptographic algorithms).

Provides new keyword that allows to mount ISO images as USB in BIOS
Menu.

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
Use common BIOS menu keywords

Rename used keyword from Tianocore based to common using
scripts/rename-keyword.sh script.

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
keys: add PAGEUP/DOWN, HOME, END as Global Variables

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
Add ansible configuration for MPL PIP device

Every new platform needs new entry in ansible-roles/hosts file

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
dasharo-security: secure-boot: reimplement tests after divided to ami/dasharo

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
lib: move more keywords to common files after tests on MPL

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
lib: ansible: add argument to ansible run KW

Provide dir for ansible runs per test suite and test on
secure-boot.robot. Remove ubuntu example. Modify lib/ansible.robot
to use argument to choose different playbook based on the test suite
name.

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
platform-config: introduce SECURE_BOOT_CAN_REMOVE_EXTERNAL_CERT variable

With AMI BIOS we are not able to reset certs enrolled from OS, to handle
such cases we added new flag to skip some tests

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
dasharo-security: secure-boot: fix tests after execution on AMI

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
variables: fixup RTE_LIST variable

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
platform-configs: mpl-pip4: run robotidy

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
dasharo-security: tpm-support: add test TPM001.004

Adds TCG2 Menu Construction keyword, return list of pairs with option
name and value, merge multiline options into one, add keyword to convert
list into OrderedDict.

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
dasharo-security: secure-boot: update tests after tests on QEMU

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
lib: dts-lib: fixes after tests on QEMU

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
lib: bios: move Exit From Current Menu to common

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
self-tests: rename leftovers after droping Tianocore from KW names

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
dasharo-security: tpm2-commands: add tests TPMCMD012.001 and 013.001-003

Use PCR 7 for policies with PCR

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
dasharo-security: tpm-support: update TPM Clear test

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
dasharo-security: tpm-support: add TPM005.001-004 tests

Check supported SHA banks from BIOS Menu

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
dasharo-security: tpm-support: add TPM006.001 and ans role for this test suite

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
dasharo-security: tpm-support: add TPM007.001 with kw to attach system

Adds Add Boot Option KW to add new Boot option for Boot Menu.

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
platform-configs: add ami/dasharo settings file for easy modification

lib: split ami/dasharo files into separate folders

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
platform-configs: add TPM_FIRMWARE_CONFIG variable

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
dasharo-security compatibility: use TPM_FIRMWARE_CONFIG var, source menus

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

Commit:
lib: fixes after running TPM Clear on AMI

Signed-off-by: Tomasz Żyjewski <tomasz.zyjewski@3mdeb.com>

TomaszAIR commented:
Results on qemu after latest rebase

```
λ robot -b file.txt -L TRACE -v ansible_config:yes -v rte_ip:127.0.0.1 -v snipeit:no -v config:qemu dasharo-security/tpm2-commands.robot
==============================================================================
Tpm2-Commands
==============================================================================
TPMCMD001.001 Check if both SHA1 and SHA256 PCRs are enabled (Ubun... | PASS |
------------------------------------------------------------------------------
TPMCMD002.001 PCRREAD Function Verification (Ubuntu 22.04) :: This... | PASS |
------------------------------------------------------------------------------
TPMCMD003.001 PCREXTEND And PCRRESET Functions (Ubuntu 22.04) :: T... | PASS |
------------------------------------------------------------------------------
TPMCMD003.002 PCREXTEND And PCRRESET Functions - locality protecti... | PASS |
------------------------------------------------------------------------------
TPMCMD004.001 PCREVENT Function (Ubuntu 22.04) :: This test aims t... | PASS |
------------------------------------------------------------------------------
TPMCMD005.001 CREATEPRIMARY Function Verification (Ubuntu 22.04) :... | PASS |
------------------------------------------------------------------------------
TPMCMD006.001 NVDEFINE and NVUNDEFINE Functions Verification (Ubun... | PASS |
------------------------------------------------------------------------------
TPMCMD007.001 CREATE Function (Ubuntu 22.04) :: This test aims to ... | PASS |
------------------------------------------------------------------------------
TPMCMD007.002 CREATELOADED Function (Ubuntu 22.04) :: This test ai... | PASS |
------------------------------------------------------------------------------
TPMCMD008.001 Signing the file (Ubuntu 22.04) :: Check whether the... | PASS |
------------------------------------------------------------------------------
TPMCMD009.001 Encryption and Decryption of the file (Ubuntu 22.04)... | SKIP |
TPM doesn't supports TPM2_EncryptDecrypt nor TPM2_EncryptDecrypt2
------------------------------------------------------------------------------
TPMCMD010.001 Hashing the file (Ubuntu 22.04) :: Check whether the... | PASS |
------------------------------------------------------------------------------
TPMCMD011.001 Performing HMAC operation on the file (Ubuntu 22.04)... | PASS |
------------------------------------------------------------------------------
TPMCMD012.001 Sealing and Unsealing the file without Policy (Ubunt... | PASS |
------------------------------------------------------------------------------
TPMCMD013.001 Sealing and Unsealing with Policy - Password Only (U... | PASS |
------------------------------------------------------------------------------
TPMCMD013.002 Sealing and Unsealing with Policy - PCR Only (Ubuntu... | PASS |
------------------------------------------------------------------------------
TPMCMD013.003 Sealing and unsealing with Policy - Password and PCR... | PASS |
------------------------------------------------------------------------------
Tpm2-Commands                                                         | PASS |
17 tests, 16 passed, 0 failed, 1 skipped
==============================================================================
Debug:   /home/tzyjewski/files/open-source-firmware-validation/file.txt
Output:  /home/tzyjewski/files/open-source-firmware-validation/output.xml
Log:     /home/tzyjewski/files/open-source-firmware-validation/log.html
Report:  /home/tzyjewski/files/open-source-firmware-validation/report.html

λ robot -b file.txt -L TRACE -v ansible_config:yes -v rte_ip:127.0.0.1 -v snipeit:no -v config:qemu dasharo-security/tpm-support.robot
==============================================================================
Tpm-Support
==============================================================================
TPM001.001 TPM Support (firmware) :: This test aims to verify that... | FAIL |
'Table not found.
' does not contain 'TPM2 log'
------------------------------------------------------------------------------
TPM001.002 TPM Support (Ubuntu 20.04) :: Check whether the TPM is ... | PASS |
------------------------------------------------------------------------------
TPM001.003 TPM Support (Windows 11) :: Check whether the TPM is in... | SKIP |
TPM001.003 not supported
------------------------------------------------------------------------------
TPM001.004 TPM Support (BIOS) :: This test aims to verify that the... | PASS |
------------------------------------------------------------------------------
TPM002.001 Verify TPM version (Ubuntu 22.04) :: This test aims to ... | PASS |
------------------------------------------------------------------------------
TPM002.002 Verify TPM version (Windows 11) :: This test aims to ve... | SKIP |
TPM002.002 not supported
------------------------------------------------------------------------------
TPM003.001 Check TPM Physical Presence Interface (firmware) :: Thi... | FAIL |
'Table not found.
' does not contain 'PPI: Pending OS request'
------------------------------------------------------------------------------
TPM003.002 Check TPM Physical Presence Interface (Ubuntu 22.04) ::... | PASS |
------------------------------------------------------------------------------
TPM003.003 Check TPM Physical Presence Interface (Windows 11) :: T... | SKIP |
TPM003.003 not supported
------------------------------------------------------------------------------
TPM004.001 Check TPM Clear procedure :: This test aims to verify w... | PASS |
------------------------------------------------------------------------------
TPM005.001 Check TPM Hash Algorithm Support SHA1 (Firmware) :: Thi... | PASS |
------------------------------------------------------------------------------
TPM005.002 Check TPM Hash Algorithm Support SHA256 (Firmware) :: T... | PASS |
------------------------------------------------------------------------------
TPM005.003 Check TPM Hash Algorithm Support SHA384 (Firmware) :: T... | PASS |
------------------------------------------------------------------------------
TPM005.004 Check TPM Hash Algorithm Support SHA512 (Firmware) :: T... | PASS |
------------------------------------------------------------------------------
TPM006.001 Encrypt and Decrypt non-rootfs partition (Ubuntu 22.04)... | PASS |
------------------------------------------------------------------------------
TPM007.001 Encrypt and Decrypt rootfs partition (Ubuntu 22.04) :: ... | PASS |
------------------------------------------------------------------------------
Tpm-Support                                                           | FAIL |
16 tests, 11 passed, 2 failed, 3 skipped
==============================================================================
Debug:   /home/tzyjewski/files/open-source-firmware-validation/file.txt
Output:  /home/tzyjewski/files/open-source-firmware-validation/output.xml
Log:     /home/tzyjewski/files/open-source-firmware-validation/log.html
Report:  /home/tzyjewski/files/open-source-firmware-validation/report.html
```

TPM001.001 and TPM003.001 fails as we started QEMU with EDK2 only so `cbmem`
does not return data.

Test executed on qemu, started with
`DRIVE_PATH=./qemu-data/ubuntu-enc.qcow2 ./scripts/ci/qemu-run.sh graphic os`

`ubuntu-enc.qcow2` is an encrypted Ubuntu image with efi partition labeled as
`ubuntu-enc` and rootfs partition labeled as `encrypted-rootfs`. Both label
names are used in test TPM007.001 scenario.

macpijan commented:
@TomaszAIR I do not understand why I see in this MR commits already present in: https://github.com/Dasharo/open-source-firmware-validation/pull/187

There seems to be a mismatch in the history. Since there is lots of duplicated code in this MR, reviewing this one would not be very productive.

TomaszAIR commented:
@macpijan I did not pushed rebased branch before. Now it should not have commits from #187

Commit:
scripts: secure-boot: generate-image: fix distro detection

lsb_release may rerturn distro with uppercase characters which will
corrupt distro detection condition, so force lowercase here.

Comparing to lsb_release, /etc/os-release ID filed never contains
uppercase characters.

Signed-off-by: Daniil Klimuk <daniil.klimuk@3mdeb.com>

Commit:
scripts: secure-boot: generate-images: delete some distros

The packages sbsigntool and faketime have not been found in the deleted
official distros repositories, so the distros are not supported.

Signed-off-by: Daniil Klimuk <daniil.klimuk@3mdeb.com>

Commit:
scripts: secure-boot: generate-image: use sudo with mkfs

Some distros require running mkfs as superuser.

Signed-off-by: Daniil Klimuk <daniil.klimuk@3mdeb.com>

Commit:
scripts: secure-boot: generate-image: dynamically determine mountpoint

udisksctl not always mounts loop devices under /run/media, sometimes it
is under /media.

Signed-off-by: Daniil Klimuk <daniil.klimuk@3mdeb.com>

Commit:
scripts: secure-boot: generate-images: generate-images: fix mounting

Try to mount image again if udisksctl has not done it automatically.

Signed-off-by: Daniil Klimuk <daniil.klimuk@3mdeb.com>

Commit:
scripts: secure-boot: generate-image: generate-images: remove loop

Remove loop device after work is done.

Signed-off-by: Daniil Klimuk <daniil.klimuk@3mdeb.com>

DaniilKl committed:
Merge pull request #368 from Dasharo/fix-issues-254-255-256

Fix issues 254 255 256
